### PR TITLE
Fix the problem of duplicate nonce; Add APIs to set m_maxRetries and m_defaultTimeout

### PR DESCRIPTION
**Pull request by matianxing1992**
https://github.com/UCLA-IRL/NAC-ABE/issues/34

### CODE REVIEW

File: src/consumer.cpp [hunk 1]
```diff
@@ -270,7 +286,11 @@ Consumer::handleTimeout(const Interest& interest, int nRetrials,
 {
   if (nRetrials > 0) {
     NDN_LOG_INFO("Timeout for: " << interest << ", retrying");
-    m_face.expressInterest(interest, dataCallback,
+    auto interestNew = const_cast<Interest&>(interest);
+    int factor = (int) std::pow(2, m_maxRetries+1-nRetrials);
```
Comment (tianyuan129) [inline src/consumer.cpp]: Can you use cast instead? Also `m_maxRetries + 1 - nRetrials`.

Comment (tianyuan129) [inline src/consumer.cpp]: Can you use cast instead? Also `m_maxRetries + 1 - nRetrials`.

File: src/consumer.cpp [hunk 2]
```diff
@@ -270,7 +286,11 @@ Consumer::handleTimeout(const Interest& interest, int nRetrials,
 {
   if (nRetrials > 0) {
     NDN_LOG_INFO("Timeout for: " << interest << ", retrying");
-    m_face.expressInterest(interest, dataCallback,
+    auto interestNew = const_cast<Interest&>(interest);
```
Comment (tianyuan129) [inline src/consumer.cpp]: Not quite understand here. Why make an Interest then cast it into another?

**Review by tianyuan129 — CHANGES_REQUESTED** [review]

File: src/consumer.cpp (hunk 2 above)
Comment (matianxing1992) [reply src/consumer.cpp]: A const interest's nonce can't be changed

Comment (tianyuan129) [reply src/consumer.cpp]: Then better make a new interest and copy over fields. This cast looks a bit strange to me tbh…solely for getting rid of const

Comment (matianxing1992) [reply src/consumer.cpp]: Yes, I will make a new interest here.

**Comment by matianxing1992** [comment]
Made some changes accordingly and submitted a new commit

**Review by tianyuan129 — APPROVED** [review]

File: src/consumer.cpp [hunk 3]
```diff
@@ -286,11 +286,12 @@ Consumer::handleTimeout(const Interest& interest, int nRetrials,
...
-    auto interestNew = const_cast<Interest&>(interest);
-    int factor = (int) std::pow(2, m_maxRetries+1-nRetrials);
-    interestNew.setInterestLifetime(ndn::time::milliseconds(m_defaultTimeout*factor));
-    interestNew.setNonce(std::nullopt);
-    m_face.expressInterest(interestNew, dataCallback,
+    Interest interestRetry(interest);
+    int factor = static_cast<int>(std::pow(2, m_maxRetries + 1 - nRetrials));
+    interestRetry.setCanBePrefix(true);
+    interestRetry.setInterestLifetime(ndn::time::milliseconds(m_defaultTimeout*factor));
+    interestRetry.setNonce(std::nullopt);
```
Comment (tianyuan129) [inline src/consumer.cpp]: Ideally, you should use refreshNonce()

**Review by tianyuan129 — CHANGES_REQUESTED** [review]

**Comment by matianxing1992** [comment]
Use refreshNonce() instead now

**Pull request merged**